python/ray/ml/utils/tensorflow_utils.py

```diff
@@ -0,0 +1,89 @@
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+from ray.data.extensions.tensor_extension import TensorDtype
+
+
+def convert_pandas_to_tf_tensor(df: pd.DataFrame) -> tf.Tensor:
+    """Convert a pandas dataframe to a TensorFlow tensor.
+
+    This function works in two steps:
+    1. Convert each dataframe column to a tensor.
+    2. Concatenate the resulting tensors along the last axis.
+
+    Arguments:
+        df: The dataframe to convert to a TensorFlow tensor. Columns must have
+            a numeric dtype, ``TensorDtype``, or object dtype. If a column has
+            an object dtype, the column must contain ``ndarray`` objects.
+
+    Returns:
+        A tensor of data type ``float32`` constructed from the dataframe.
+
+    Raises:
+        ValueError: if a column has an invalid dtype.
+        ValueError: if the columns can't be combined into a single tensor.
+
+    Examples:
+        >>> import pandas as pd
+        >>> from ray.ml.utils.tensorflow_utils improt convert_pandas_to_tf_tensor
+        >>>
+        >>> df = pd.DataFrame({"X1": [1, 2, 3], "X2": [4, 5, 6]})
+        >>> convert_pandas_to_tf_tensor(df[["X1"]]).shape
+        TensorShape([3, 1])
+        >>> convert_pandas_to_tf_tensor(df[["X1", "X2"]]).shape
+        TensorShape([3, 2])
+
+        >>> from ray.data.extensions import TensorArray
+        >>>
+        >>> df = pd.DataFrame({"image": TensorArray(np.zeros((4, 3, 32, 32)))})
+        >>> convert_pandas_to_tf_tensor(df).shape
+        TensorShape([4, 3, 32, 32])
+    """
+    def is_valid_dtype(series) -> bool:
+        if pd.api.types.is_numeric_dtype(series) or isinstance(series, TensorDtype):
+            return True
+
+        is_ndarray = series.map(lambda obj: isinstance(obj, np.ndarray))
+        return all(is_ndarray)
+
+    for column, series in df.iteritems():
+        if not is_valid_dtype(series):
+            raise ValueError(
+                f"Expected column {column} to have numeric dtype, "
+                "`TensorDtype`, or object dtype with `ndarray` elements. "
+                f"Instead, received dtype {series.dtype}."
+            )
```

Comment:
TODO: Input validation should probably happen at the public API layer (e.g., `Dataset.to_tf`)

Reply:
Also, because of duck typing, I'm not sure if this validation is needed at all

Reply:
I don't think that we should do this validation, since TensorFlow will do their own validation, and if those error messages end up being opaque, we can always try + except + pattern match + raise our own more informative exception. This should also prevent us from having to do extra full column scans in the ndarray object case, and is a general EAFP vs. LBYL, with the former being more idiomatic Python: https://mathspp.com/blog/pydonts/eafp-and-lbyl-coding-styles

Reply:
Fixed